cmd/snap-confine/udev-support.c

```diff
@@ -18,290 +18,355 @@
-					       "%s", security_tag);
-	for (size_t i = 0; i < udev_s->tagname_len; i++)
-		if (udev_s->tagname[i] == '.')
-			udev_s->tagname[i] = '_';
+	/* Allow access to /dev/nvidia0 through /dev/nvidia254 */
+	for (unsigned nv_minor = 0; nv_minor < 255; nv_minor++) {
+		char nv_path[15] = { 0 };	// /dev/nvidiaXXX
+		sc_must_snprintf(nv_path, sizeof(nv_path), "/dev/nvidia%u",
+				 nv_minor);
 
-	udev_s->udev = udev_new();
-	if (udev_s->udev == NULL)
-		die("udev_new failed");
+		/* Stop trying to find devices after one is not found. In this manner,
+		 * we'll add /dev/nvidia0 and /dev/nvidia1 but stop trying to find
+		 * nvidia3 - nvidia254 if nvidia2 is not found. */
+		if (stat(nv_path, &sbuf) < 0) {
+			break;
+		}
+		dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			minor(sbuf.st_rdev));
+	}
 
-	udev_s->devices = udev_enumerate_new(udev_s->udev);
-	if (udev_s->devices == NULL)
-		die("udev_enumerate_new failed");
+	if (stat("/dev/nvidiactl", &sbuf) == 0) {
+		dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			minor(sbuf.st_rdev));
+	}
+	if (stat("/dev/nvidia-uvm", &sbuf) == 0) {
+		dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			minor(sbuf.st_rdev));
+	}
+	if (stat("/dev/nvidia-modeset", &sbuf) == 0) {
+		dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			minor(sbuf.st_rdev));
+	}
+}
 
-	if (udev_enumerate_add_match_tag(udev_s->devices, udev_s->tagname) < 0)
-		die("udev_enumerate_add_match_tag");
+/**
+ * Allow access to /dev/uhid.
+ *
+ * Currently /dev/uhid isn't represented in sysfs, so add it to the device
+ * cgroup if it exists and let AppArmor handle the mediation.
+ **/
+static void sc_udev_allow_uhid(int devices_allow_fd)
+{
+	struct stat sbuf;
 
-	if (udev_enumerate_scan_devices(udev_s->devices) < 0)
-		die("udev_enumerate_scan failed");
+	if (stat("/dev/uhid", &sbuf) == 0) {
+		dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			minor(sbuf.st_rdev));
+	}
+}
 
-	udev_s->assigned = udev_enumerate_get_list_entry(udev_s->devices);
-	if (udev_s->assigned == NULL)
-		rc = -1;
+/**
+ * Allow access to assigned devices.
+ *
+ * The snapd udev security backend uses udev rules to tag matching devices with
+ * tags corresponding to snap applications. Here we interrogate udev and allow
+ * access to all assigned devices.
+ **/
+static void sc_udev_allow_assigned(int devices_allow_fd, struct udev *udev,
+				   struct udev_list_entry *assigned)
+{
+	for (struct udev_list_entry * entry = assigned; entry != NULL;
+	     entry = udev_list_entry_get_next(entry)) {
+		const char *path = udev_list_entry_get_name(entry);
+		if (path == NULL) {
+			die("udev_list_entry_get_name failed");
+		}
+		struct udev_device *device =
+		    udev_device_new_from_syspath(udev, path);
+		if (device == NULL) {
+			die("cannot find device from syspath %s", path);
+		}
+		dev_t devnum = udev_device_get_devnum(device);
+		char type_c = strstr(path, "/block/") != NULL ? 'b' : 'c';
```

Comment:
Maybe we should use `udev_device_get_devnode()` and then `stat()` the resulting path to get the type or check whether subsystem is `"block"`?

Reply:
Done in 02b48d5645b0e4169dead186cb1b2a0a7e816b1e